### PR TITLE
feat: self service change user password

### PR DESCRIPTION
**Pull request by BLasan**
## Description
Modal for password changing of the users.

## Related issues and discussion
#669 

## Screenshots, if any
Screenshot from 2020-05-16 13-39-20

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

File: src/app/shared/change-password-dialog/change-password-dialog.component.html [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<h1 mat-dialog-title>Change Password</h1>
+<div mat-dialog-content>
```
Comment (karantakalkar) [inline src/app/shared/change-password-dialog/change-password-dialog.component.html]: can you please delineate the HTML.

**Comment by BLasan** [comment]
@abhaychawla  Please review the PR

**Comment by karantakalkar** [comment]
@BLasan so change password dialog is being re used here?

File: src/app/self-service/users/user.service.ts [hunk 2]
```diff
@@ -54,4 +55,36 @@ export class UserService {
...
+        if (controlToCompare) {
+            const subscription: Subscription = controlToCompare.valueChanges.subscribe(() => {
+                c.updateValueAndValidity();
+                subscription.unsubscribe();
+            });
+        }
+        return controlToCompare && controlToCompare.value !== c.value ? {'notequal': true} : null;
+    };
+  }
+
```
Comment (karantakalkar) [inline src/app/self-service/users/user.service.ts:87]: I think we should use a password service for both these functions

File: src/app/self-service/users/user.service.ts [hunk 3]
```diff
@@ -54,4 +55,36 @@ export class UserService {
...
+
+
```
Comment (karantakalkar) [inline src/app/self-service/users/user.service.ts:89]: take care of extra lines

File: src/app/self-service/users/view-user/view-user.component.ts [hunk 4]
```diff
@@ -30,8 +40,23 @@ export class ViewUserComponent implements OnInit {
...
+   */
+  changeUserPassword() {
+    const changeUserPasswordDialogRef = this.dialog.open(ChangePasswordDialogComponent, {
+      data: { password: `user ${this.user.id}` }
+    });
+
+    changeUserPasswordDialogRef.afterClosed().subscribe((response: any) => {
+      if (response.password && response.repeatPassword) {
+        const password = response.password;
+        const repeat_password = response.repeatPassword;
```
Comment (karantakalkar) [inline src/app/self-service/users/view-user/view-user.component.ts:54]: camel casing only

File: src/app/shared/change-password-dialog/change-password-dialog.component.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import { Component, OnInit, Inject } from '@angular/core';
+import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material';
+import { FormBuilder, Validators } from '@angular/forms';
+import { UserService } from 'app/self-service/users/user.service';
+@Component({
```
Comment (karantakalkar) [inline src/app/shared/change-password-dialog/change-password-dialog.component.ts:5]: comments and spacing

File: src/app/shared/change-password-dialog/change-password-dialog.component.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  selector: 'mifosx-change-password-dialog',
+  templateUrl: './change-password-dialog.component.html',
+  styleUrls: ['./change-password-dialog.component.scss']
+})
+export class ChangePasswordDialogComponent implements OnInit {
+
+  /**
+   * Change Password Form
+   */
```
Comment (karantakalkar) [inline src/app/shared/change-password-dialog/change-password-dialog.component.ts:14]: single line jsdoc

File: src/app/shared/change-password-dialog/change-password-dialog.component.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  changePasswordForm: any;
+
+  /**
+   * @param {MatDialogRef} dialogRef Component reference to dialog.
+   * @param {any} data Provides any data.
+   */
+  constructor(public dialogRef: MatDialogRef<ChangePasswordDialogComponent>,
+    @Inject(MAT_DIALOG_DATA) public data: any,
```
Comment (karantakalkar) [inline src/app/shared/change-password-dialog/change-password-dialog.component.ts:22]: indent

File: src/app/users/users.service.ts [hunk 8]
```diff
@@ -48,6 +49,36 @@ export class UsersService {
...
+  /**
+   * Change User Password
+   * @param userId User Id of users
+   * @param password New Password of the user
+   * @returns {Observable<any>}
+   */
+  changePassword(userId: string, passwordObj: any) {
+    return this.http.put(`/users/${userId}`, passwordObj);
+  }
+
```
Comment (karantakalkar) [inline src/app/users/users.service.ts:81]: why are the same functions in 2 services?

File: src/app/users/view-user/view-user.component.ts [hunk 9]
```diff
@@ -58,4 +59,24 @@ export class ViewUserComponent implements OnInit {
...
 
+
+  /**
+   * Change Password of the Users
+   */
+  changeUserPassword() {
+    const changeUserPasswordDialogRef = this.dialog.open(ChangePasswordDialogComponent, {
+      data: { password: `user ${this.userData.id}` }
+    });
+
```
Comment (karantakalkar) [inline src/app/users/view-user/view-user.component.ts:70]:
```suggestion
```

File: src/app/self-service/users/user.service.ts (hunk 2 above)
Comment (BLasan) [reply src/app/self-service/users/user.service.ts:87]: Yes, In #660 I put the function in component.ts . Do we have to create a new service file and put it for the reusability?

Comment (karantakalkar) [reply src/app/self-service/users/user.service.ts:87]: yes

**Comment by BLasan** [comment]
> @BLasan so change password dialog is being re used here?

Yes

File: src/app/shared/change-password-dialog/change-password-dialog.component.ts (hunk 5 above)
Comment (BLasan) [reply src/app/shared/change-password-dialog/change-password-dialog.component.ts:12]: Done

File: src/app/users/users.service.ts (hunk 8 above)
Comment (BLasan) [reply src/app/users/users.service.ts:61]: What is the other service this function being repeated?

File: src/app/users/view-user/view-user.component.ts (hunk 9 above)
Comment (BLasan) [reply src/app/users/view-user/view-user.component.ts:70]: Done

**Comment by karantakalkar** [comment]
@BLasan you will need to rebase once #660 is merged you can synchronize this with code of #660 however.

**Comment by BLasan** [comment]
> @BLasan you will need to rebase once #660 is merged you can synchronize this with code of #660 however.

Okay :+1:

**Comment by karantakalkar** [comment]
@BLasan synchronize with change user password and remove MCs

**Comment by BLasan** [comment]
> @BLasan synchronize with change user password and remove MCs

I'll do it now

**Comment by BLasan** [comment]
@karantakalkar  Done. Please have a look

**Comment by karantakalkar** [comment]
tag @muskankhedia as well

File: src/app/shared/change-password-dialog/password-check.service.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
+/** Angular Imports */
```
Comment (karantakalkar) [inline src/app/shared/change-password-dialog/password-check.service.ts:1]: shouldn't this service be deleted? the function is already there in dialog.component

File: src/app/shared/shared.module.ts [hunk 11]
```diff
@@ -16,7 +16,6 @@ import { ChangePasswordDialogComponent } from './change-password-dialog/change-p
 /** Custom Modules */
 import { IconsModule } from './icons.module';
 import { MaterialModule } from './material.module';
-
```
Comment (karantakalkar) [inline src/app/shared/shared.module.ts:19]: restore the line

File: src/app/users/users.service.ts [hunk 12]
```diff
@@ -3,7 +3,8 @@ import { Injectable } from '@angular/core';
 import { HttpClient, HttpParams } from '@angular/common/http';
 
 /** rxjs Imports */
-import { Observable } from 'rxjs';
+import { Observable, Subscription } from 'rxjs';
+import { AbstractControl, ValidatorFn, ValidationErrors } from '@angular/forms';
```
Comment (karantakalkar) [inline src/app/users/users.service.ts:7]:
```suggestion
```

**Comment by BLasan** [comment]
@karantakalkar  @muskankhedia  Please have a look

File: src/app/self-service/users/user.service.ts [hunk 13]
```diff
@@ -54,4 +53,14 @@ export class UserService {
...
   }
 
+  /**
+   * Change User Password
+   * @param userId User Id of users
+   * @param password New Password of the user
+   * @returns {Observable<any>}
+   */
+  changePassword(userId: string, passwordObj: any) {
+    return this.http.put(`/users/${userId}`, passwordObj);
```
Comment (muskankhedia) [inline src/app/self-service/users/user.service.ts:63]: Please follow the documentation for the API details. [Link](https://demo.mifos.io/api-docs/apiLive.htm#selfuser) for the self user service.

Comment (BLasan) [reply src/app/self-service/users/user.service.ts]: Resolved

**Comment by BLasan** [comment]
@karantakalkar @muskankhedia  Done. Please review

**Comment by karantakalkar** [comment]
@BLasan change html of change password dialog to this https://gist.github.com/karantakalkar/b50521539c711d90b444c1280b35dbb1

File: src/app/self-service/users/user.service.ts [hunk 14]
```diff
@@ -26,7 +26,6 @@ export class UserService {
 
   /**
    * Gets all the self service users.
-   *
```
Comment (karantakalkar) [inline src/app/self-service/users/user.service.ts:29]: remove this change

**Comment by BLasan** [comment]
> @BLasan change html of change password dialog to this https://gist.github.com/karantakalkar/b50521539c711d90b444c1280b35dbb1

Done

File: src/app/shared/change-password-dialog/change-password-dialog.component.html [hunk 15]
```diff
@@ -47,4 +43,4 @@ <h1 mat-dialog-title>Change Password</h1>
 <mat-dialog-actions align="center">
   <button mat-raised-button mat-dialog-close>Cancel</button>
   <button mat-raised-button color="warn" [disabled]="!changePasswordForm.valid" [mat-dialog-close]="{password:changePasswordForm.get('password').value,repeatPassword:changePasswordForm.get('repeatPassword').value}">Confirm</button>
-</mat-dialog-actions>
+</mat-dialog-actions>
```
Comment (muskankhedia) [inline src/app/shared/change-password-dialog/change-password-dialog.component.html:46]: please add a line at the end of line

**Comment by BLasan** [comment]
@muskankhedia @karantakalkar  Done. Please take a look

File: src/app/self-service/users/user.service.ts [hunk 16]
```diff
@@ -54,4 +54,14 @@ export class UserService {
     return this.http.disableApiPrefix().get('/assets/mock/user.mock.json');
   }
 
+  /**
+   * Change User Password
+   * @param userId User Id of users
+   * @param password New Password of the user
+   * @returns {Observable<any>}
+   */
+  changePassword(userId: string, passwordObj: any) {
```
Comment (karantakalkar) [inline src/app/self-service/users/user.service.ts:63]: @BLasan just add a comment here `// TODO: update endpoint once API available`, then we'll merge

Comment (BLasan) [reply src/app/self-service/users/user.service.ts:63]: Ok, I'll update that now

Comment (BLasan) [reply src/app/self-service/users/user.service.ts:65]: Done. Please review

**Comment by karantakalkar** [comment]
@BLasan fix lin errors

**Comment by BLasan** [comment]
> @BLasan fix lin errors

Sorry, I'll fix it

**Comment by BLasan** [comment]
@karantakalkar @muskankhedia  Please review

**Pull request merged**